p2p/discover/udp_quorum.go

```diff
@@ -0,0 +1,479 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package discover
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/netutil"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// RPC request structures
+type (
+	pingQuorum struct {
+		Version    uint
+		From, To   rpcEndpoint
+		Expiration uint64
+		Hostname   string
+		// Ignore additional fields (for forward compatibility).
+		Rest []rlp.RawValue `rlp:"tail"`
+	}
+
+	// pong is the reply to ping.
+	pongQuorum struct {
+		// This field should mirror the UDP envelope address
+		// of the ping packet, which provides a way to discover the
+		// the external address (after NAT).
+		To rpcEndpoint
+
+		ReplyTok   []byte // This contains the hash of the ping packet.
+		Expiration uint64 // Absolute timestamp at which the packet becomes invalid.
+		// Ignore additional fields (for forward compatibility).
+
+		Rest []rlp.RawValue `rlp:"tail"`
+	}
+
+	// findnode is a query for nodes close to the given target.
+	findnodeQuorum struct {
+		Target     encPubkey
+		Expiration uint64
+		// Ignore additional fields (for forward compatibility).
+
+		IsQuorum   bool
```

Comment:
Why do we need this field here?

Same for for `neighborsQuorum`

Reply:
I considered this a cheap way to differentiate the packet from a DiscV4 packet. A DiscV4 packet will fail decoding when checked in the Quorum DiscV4 decoder, and get passed on appropriately.